src/p2p/swarm.rs

```diff
@@ -335,36 +347,29 @@ impl NetworkBehaviour for SwarmApi {
 
     fn inject_event(&mut self, _peer_id: PeerId, _connection: ConnectionId, _event: void::Void) {}
 
-    fn inject_dial_failure(&mut self, peer_id: &PeerId) {
-        trace!("inject_dial_failure: {}", peer_id);
-        if self.pending_addresses.contains_key(peer_id) {
-            // it is possible that these addresses have not been tried yet; they will be asked
-            // for soon.
-            self.events
-                .push_back(swarm::NetworkBehaviourAction::DialPeer {
-                    peer_id: *peer_id,
-                    condition: DialPeerCondition::NotDialing,
-                });
-        }
-
-        // this should not be executed once, but probably will be in case unsupported addresses or something
-        // surprising happens.
-        for failed in self.pending_connections.remove(peer_id).unwrap_or_default() {
-            self.connect_registry
-                .finish_subscription(failed.into(), Err("addresses exhausted".into()));
-        }
-    }
-
-    fn inject_addr_reach_failure(
+    fn inject_dial_failure(
         &mut self,
-        peer_id: Option<&PeerId>,
-        addr: &Multiaddr,
-        error: &dyn std::error::Error,
+        peer_id: Option<PeerId>,
+        _handler: Self::ConnectionHandler,
+        error: &DialError,
     ) {
-        trace!("inject_addr_reach_failure {} {}", addr, error);
+        trace!("inject_dial_failure: {:?} ({})", peer_id, error);
+        if let Some(peer_id) = peer_id {
+            if self.pending_addresses.contains_key(&peer_id) {
+                // it is possible that these addresses have not been tried yet; they will be asked
+                // for soon.
+                let handler = self.new_handler();
+                self.events.push_back(swarm::NetworkBehaviourAction::Dial {
```

Comment:
All right ... Looking at the networkbehaviour, I think this is how it used to work:

1. dial event is given to swarm
2. swarm collects all of the known addresses with `NetworkBehaviour::addresses_of_peer`, dials them somehow
3. for each of the dials, we used to get a signal that this multiaddr failed and we would signal that future as failure
4. there used to be another trait method for having exhausted the addresses, when we'd know to launch a new dial if we'd still have addresses

While writing this @mxinden [replied](https://github.com/rs-ipfs/rust-ipfs/pull/499#issuecomment-1073724116). Oki yeah it would appear the failures have now moved within the dialerror, which I did not expect, AND there is only one "notification" for all of the attempts.

However I think the idea with the original impl was that since there would be one gathering of addresses for the peer (`NetworkBehaviour::addresses_of_peer`) per dial events (caused either by "swarm_api" or by any other place) it would be possible to add new addresses during a dial attempt, and those would be noticed at (4) in the above ordered list, and thus get dialed afterwards.

Reply:
I think for the next steps would be to gather the failed addresses from the error (if found), then continue dialing to the remaining addresses, if any.